feat: Add support for MongoDB 7

coratgerl opened the pull request:
## Pull Request

This PR add the elements to support the MongoDB v7

Closes: https://github.com/parse-community/parse-server/issues/8713

## Approach
I added the test script in package.json, I ran the script and no one test failed.

## Tasks

- [x] Add tests

coratgerl committed:
feat: add mongodb v7 support

coratgerl committed:
feat: update readme

coratgerl committed:
feat: update minor also in ci

mtrezza reviewed (changes requested):
Looks good, thanks for picking this up! Parse Server tests are passing. For full MongoDB 7 support, the MongoDB driver needs to be upgraded as well to the latest version. The current version `"mongodb": "4.10.0"` is only partially compatible with MongoDB 7, see [docs](https://www.mongodb.com/docs/drivers/node/current/compatibility/).

coratgerl committed:
wip

coratgerl committed:
fix: tests

coratgerl committed:
feat: update mongo to fix memory leaks

coratgerl committed:
refactor: remove useless param

coratgerl commented:
I updated the version of mongodb to 5.9 to keep the support for node 14.21. I also fix the tests that failed suits to this upgrade. Finally I updated the mongodb-runner to a more recent version :). In local test with mongo v7 works with node 18 and node 14 :)@mtrezza

coratgerl committed:
fix: lint

coratgerl committed:
fix: replica set

coratgerl committed:
fix: typo

coratgerl committed:
fix: package json

coratgerl committed:
fix: ci

coratgerl committed:
fix: weird lock move

coratgerl committed:
fix: lock with good npm version

coratgerl committed:
fix: lock again 🫣

coratgerl committed:
fix: tests

coratgerl committed:
fix: tests < 4.4

mtrezza commented on package.json:
```diff
@@ -118,18 +118,19 @@
     "build": "babel src/ -d lib/ --copy-files",
     "watch": "babel --watch src/ -d lib/ --copy-files",
-    "test:mongodb:runnerstart": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start",
+    "test:mongodb:runnerstart": "cross-env mongodb-runner stop --all && mongodb-runner start -t standalone --version $npm_config_dbversion -- --port 27017",
```
- Keep the env var fallback `MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion}`.
- Remove the `mongodb-runner stop --all` from the start command, it's already in `posttest`.

mtrezza commented on package.json:
```diff
@@ -118,18 +118,19 @@
     "testonly": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} TESTING=1 jasmine",
     "test": "npm run testonly",
-    "posttest": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner stop",
+    "posttest": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} mongodb-runner stop -t standalone --all",
```
I believe this can be reduced to `mongodb-runner stop --all`?

mtrezza commented on package.json:
```diff
@@ -118,18 +118,19 @@
-    "posttest:mongodb": "mongodb-runner stop",
-    "pretest": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start",
+    "test:mongodb:7.0.1": "npm run test:mongodb --dbversion=7.0.1",
+    "posttest:mongodb": "mongodb-runner stop --all",
```
Is `posttest:mongodb` ever called? If not, we can remove the script.

mtrezza commented on spec/vulnerabilities.spec.js:
Why has this been removed? It's an important check in this vulnerability test, so before we can remove that we need to understand why.

coratgerl replied:
Yes of course, I removed this because in the migration guide of the bson v5 they say this (see screen bellow). Now it is always return Code instances
<img width="1019" alt="image" src="https://github.com/parse-community/parse-server/assets/73360179/ee7d0262-8d5a-42a5-91a1-ce1100f78767">

coratgerl replied:
https://github.com/mongodb/js-bson/blob/main/docs/upgrade-to-v5.md

mtrezza replied:
The test focuses on "BSON type code data"; if the `_bsontype: 'Code',` is removed, how would the MongoDB driver (or DB) know that this is code?

coratgerl replied:
Hum, I have a doubt but they this ` The evalFunctions option, when enabled, would return BSON Code typed values as eval-ed JavaScript functions`. For me the evalFunctions always return a Code type. If I explicetely put bsontype to Code we have an Internal error return by mongo 🤔

mtrezza replied:
What does the MongoDB driver changelog say?

coratgerl replied:
They nothing about this juste they upgrade to bson v5. This information is in the migration guide of bson v5

mtrezza replied:
> we have an Internal error return by mongo 

Could you post the error?

coratgerl replied:
The only information we have is an error 500 : "Internal server error"

mtrezza replied:
Could you set a breakpoint inside Parse Server to get more details? It's likely an error thrown by the MongoDB driver that gets abstracted by Parse Server.

mtrezza replied:
See this discussion for maybe some insight into this: https://github.com/parse-community/parse-server/pull/8560#discussion_r1204953867

mtrezza commented on spec/ParseQuery.hint.spec.js:
```diff
@@ -178,16 +178,17 @@ describe_only_db('mongo')('Parse.Query hint', () => {
-    expect(queryPlanner.winningPlan.stage).toBe('FETCH');
-    expect(queryPlanner.winningPlan.inputStage.keyPattern).toEqual({ _id: 1 });
+    queryPlanner = result[0].queryPlanner;
+
```
```suggestion
```

mtrezza commented on src/Adapters/Storage/Mongo/MongoCollection.js:
```diff
@@ -160,18 +160,10 @@ export default class MongoCollection {
-          }
-        }
-      );
+    return this._mongoCollection.createIndex(indexRequest, {
```
Does this return a promise? I assume the reason why this was wrapped in a promise previously was that it wasn't returning a promise?

coratgerl replied:
Yes in the mongo v4 version, they use callback so we needed Promise, but in the mongo v5 the callback is removed so we don't need anymore new Promise :)

mtrezza commented on src/Adapters/Storage/Mongo/MongoStorageAdapter.js:
```diff
@@ -31,6 +31,8 @@ const storageAdapterAllCollections = mongoAdapter => {
     .connect()
     .then(() => mongoAdapter.database.collections())
     .then(collections => {
+      if (!collections || !collections.length) return [];
```
Why is this needed? Just to make sure we don't introduce a breaking change here.

coratgerl replied:
It's needed because in some of tests due to invalid id, .collections() don't return an array :). So collections.filter crash because it isn't an array :)

mtrezza replied:
> collections() don't return an array

When is that the case, for which test? Is that a behavior change in the MongoDB driver?

coratgerl replied:
The test are in apple auth and facebook auth. Like 'should use algorithm from key header to verify id_token'.

The problem is thats test spyOn utils.promisify, but now database.collections seems to use it so it's mock this calls also. I search a better solution to mock differently but I don't found anything. With jasmine we can't mock only for onetime, we can't also export the function of facebook adapter and mock it because it's not the same instance. I have no idea. Do you have a proposition @mtrezza  ?

mtrezza replied:
`spyOn(util, 'promisify')` looks like bad practice; this general helper function can be used anywhere and break the test, as it does here. An alternative would be to look at what it's supposed to mock and mock on a higher level. What should be mocked, the `authUtils`?

coratgerl replied:
We need to mock this function : 
```javascript
const getFacebookKeyByKeyId = async (keyId, cacheMaxEntries, cacheMaxAge) => {
  const client = jwksClient({
    jwksUri: `${TOKEN_ISSUER}/.well-known/oauth/openid/jwks/`,
    cache: true,
    cacheMaxEntries,
    cacheMaxAge,
  });

  const asyncGetSigningKeyFunction = util.promisify(client.getSigningKey);

  let key;
  try {
    key = await asyncGetSigningKeyFunction(keyId);
  } catch (error) {
    throw new Parse.Error(
      Parse.Error.OBJECT_NOT_FOUND,
      `Unable to find matching key for Key ID: ${keyId}`
    );
  }
  return key;
};
```

mtrezza replied:
I think we have 3 options:
- a) mock `getFacebookKeyByKeyId` -> not preferred, as we'd be not testing the code in this function
- b) mock `client.getSigningKey`
- c) mock `asyncGetSigningKeyFunction ` -> requires refactoring the code to provide a way to mock it

I'd go with (b).

coratgerl replied:
How can I mock `client.getSigningKey` with jasmine (I am not familiar with it). Moreover the client is created in the function so how can we mock the same instance ?

mtrezza replied:
See how `spyOn` and `spyOnProperty` is used in the tests

coratgerl replied:
I looked but in this case the problem is that the instant of client in the function is different that every function that we mock. So the mocked function is never called. I don't know if in Jasmine it's possible to mock an internal variable of a function

mtrezza replied:
Could you tell me what are the tests that are failing in which we need to mock this, then I'll take a look

coratgerl replied:
Almost a big part of the facebook test and apple test in AuthenticationAdapter.spec.js

mtrezza replied:
I mean which line in the test specifically?

mtrezza replied:
@coratgerl I hope you didn't get discouraged due to the scope of this PR; it would be great if we could get this ready for merge.

coratgerl replied:
No of course @mtrezza just so busy last days ahah.

This test for example :) 'should use algorithm from key header to verify id_token'. Then we have a bad value in collections here : https://github.com/parse-community/parse-server/blob/d2fda507d9420ec8fc4b32411b832c53419e2cd4/src/Adapters/Storage/Mongo/MongoStorageAdapter.js#L34

coratgerl replied:
Do you have a solution for this problem @mtrezza  ? It is the only things missing for merge I think

mtrezza replied:
I just didn't find time yet to look into it, but I'm sure it can be solved

mtrezza replied:
@abhisheksharma010 do you want to look into this? Since you asked to work on this issue

coratgerl replied:
@mtrezza I was able to free up a little time on it, I found a solution that seems to fix the problem :)

mtrezza commented on src/Adapters/Storage/Mongo/MongoStorageAdapter.js:
```diff
@@ -687,13 +689,10 @@ export class MongoStorageAdapter implements StorageAdapter {
+      .then(collection =>
+        collection._mongoCollection
+          .createIndex(indexCreationRequest, indexOptions)
+          .catch(error => this.handleError(error))
```
Why ` .catch(error => this.handleError(error))` if there is already a `handleError` after this?

coratgerl replied:
Yes you right :), i revert this change :)

mtrezza committed:
Update spec/ParseQuery.hint.spec.js

Signed-off-by: Manuel <5673677+mtrezza@users.noreply.github.com>

coratgerl committed:
fix: 4.4 tests

mtrezza commented:
Nice one, thank's for also updating the mongodb-runner, that is overdue.

coratgerl committed:
Merge branch 'feat-support-mongo-v7' of github.com:coratgerl/parse-server into feat-support-mongo-v7

# Conflicts:
#	spec/ParseQuery.hint.spec.js

mtrezza committed:
Merge branch 'alpha' into feat-support-mongo-v7

mtrezza committed:
Merge branch 'alpha' into feat-support-mongo-v7

coratgerl committed:
fix: feedbacks

coratgerl committed:
Merge branch 'feat-support-mongo-v7' of github.com:coratgerl/parse-server into feat-support-mongo-v7

coratgerl commented:
Thanks for you feedbacks, I fixed them, and make my feedbacks on others :)

mtrezza commented on package.json:
```diff
@@ -118,18 +118,18 @@
     "build": "babel src/ -d lib/ --copy-files",
     "watch": "babel --watch src/ -d lib/ --copy-files",
-    "test:mongodb:runnerstart": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start",
+    "test:mongodb:runnerstart": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start -t standalone --version $npm_config_dbversion -- --port 27017",
```
We are getting the topology via `MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} `; I believe this should be passed to the runner instead of a fixed `standalone`?

mtrezza replied:
Why was this resolved?

mtrezza commented on package.json:
```diff
@@ -118,18 +118,18 @@
-    "posttest:mongodb": "mongodb-runner stop",
-    "pretest": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start",
+    "test:mongodb:7.0.1": "npm run test:mongodb --dbversion=7.0.1",
+    "pretest": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start -- --port 27017",
```
This seens to be missing the topology to be passed to the runner?

mtrezza replied:
This was marked as resolved, but it still is missing the topology and version params to be passed.

coratgerl replied:
Lol you are so fast, I worked in local, it's resolved in my local not yet pushed eheh

mtrezza replied:
Please only resolve after the commit has been pushed to this branch, otherwise it's unclear how it has been resolved.

mtrezza commented on package.json:
```diff
@@ -118,18 +118,18 @@
     "testonly": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} TESTING=1 jasmine",
     "test": "npm run testonly",
-    "posttest": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner stop",
+    "posttest": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=5.3.2} mongodb-runner stop --all",
```
I don't think we need to know the mongodb version here?

```suggestion
    "posttest": "mongodb-runner stop --all",
```

mtrezza replied:
Also here; why was this resolved?

coratgerl replied:
Lol you are so fast, I worked in local, it's resolved in my local not yet pushed eheh

coratgerl committed:
fix: package

mtrezza commented on package.json:
```diff
@@ -118,18 +118,18 @@
     "build": "babel src/ -d lib/ --copy-files",
     "watch": "babel --watch src/ -d lib/ --copy-files",
-    "test:mongodb:runnerstart": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start",
+    "test:mongodb:runnerstart": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start -t ${MONGODB_TOPOLOGY} --version $npm_config_dbversion -- --port 27017",
```
The purpose of these env vars and fallbacks is to use them in the runner:

```suggestion
    "test:mongodb:runnerstart": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start -t ${MONGODB_TOPOLOGY} --version ${MONGODB_VERSION} -- --port 27017",
```

coratgerl committed:
fix: feedbacks

mtrezza commented:
It seems that since you changed the way the query plan results are tested, the tests are now failing for [MongoDB 4.2](https://github.com/parse-community/parse-server/actions/runs/6289806017/job/17076646002?pr=8761). You'd need to introduce an additional test for MongoDB 4.2. Look at the `it_only_mongodb_version` tests to see how this is done with other tests using range operators.

coratgerl committed:
fix: bson test

kantorcodes commented:
Would be fantastic to see this merged :)

coratgerl committed:
fix: tests

coratgerl committed:
Merge branch 'alpha' into feat-support-mongo-v7

# Conflicts:
#	package-lock.json

coratgerl committed:
fix: remove fdescribe

coratgerl committed:
fix: try to fix 4.2 tests

coratgerl committed:
fix: try to fix <4.4

coratgerl committed:
fix: display

coratgerl committed:
fix: display

coratgerl committed:
fix: 4.2 tests

coratgerl committed:
fix: first part of test

coratgerl committed:
fix: second part of test

coratgerl committed:
fix: all tests

coratgerl committed:
fix: remove change to package json for tests purpose

coratgerl commented:
@mtrezza After 1 hour to tryhard to make 4.2 tests passed (I can't run 4.2 tests on my compute, no mongodb-runner available for this version with my arch) all tests passed. For me it's good !

mtrezza commented:
Great! Let me review this...

mtrezza committed:
Merge branch 'alpha' into feat-support-mongo-v7

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -1517,14 +1517,16 @@ describe('apple signin auth adapter', () => {
     expect(result).toEqual(fakeClaim);
     expect(jwt.verify.calls.first().args[2].algorithms).toEqual(fakeDecodedToken.header.alg);
+
+    spy.and.callThrough();
```
I still believe we should not spy on a general method like `promisify`. These test can easily break when some unrelated code is changed in Parse Server. This could make the bugs difficult to identify and cost a lot of time in the future to investigate.

coratgerl replied:
I know, but I don't find a better solution. It's not for lack of looking :(

mtrezza replied:
I know you already invested some time into this. Let me see if I can find another solution, otherwise we merge as is.

mtrezza replied:
Fixed in https://github.com/parse-community/parse-server/pull/8761/commits/e4502a7f3e06e27566460a66a452b2c9b0eb6481

mtrezza commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -457,6 +457,14 @@ const parseObjectKeyValueToMongoObjectKeyValue = (restKey, restValue, schema) =>
   }
   value = mapValues(restValue, transformInteriorValue);
+
+  if (
```
Could you explain this part? This seems to be hacking the MongoDB driver?

coratgerl replied:
I follow the line used by @dblythy here https://github.com/parse-community/parse-server/pull/8560#discussion_r1204953867

mtrezza replied:
Got it. The reason the PR wasn't merged was because it wasn't fully clear if that was a breaking change. As @dblythy noted that some types are not supported and may need to be added in the future. I haven't looked deeper into this. @dblythy could you jump in here and maybe help us to make a decision whether this is safe to merge?

mtrezza replied:
Not sure what the implication of this is, https://github.com/parse-community/parse-server/pull/8560#discussion_r1222366763 is still unclear to me. But since this is an open topic for months without moving ahead, I believe we should take the risk and merge it into Parse Server 7. Since Parse Server 7 will be a major release, developers are supposed to test their app thoroughly before upgrading anyway. So if this really causes a braking change as I understand from https://github.com/parse-community/parse-server/pull/8560#discussion_r1222366763 is possible, then at least it will be with a major release, albeit not documented in the changelog.

mtrezza committed:
fix spy on promisify

mtrezza commented:
@coratgerl I've resolved the 2 open conversations; could you resolve the conflict, so we can go ahead and merge?

coratgerl committed:
Merge branch 'alpha' into feat-support-mongo-v7

# Conflicts:
#	package-lock.json

coratgerl commented:
Thanks you @mtrezza, the conflicts are resolved, we can go ahead !

coratgerl committed:
fix: lint

coratgerl committed:
fix: npm

Moumouls commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -457,6 +457,14 @@ const parseObjectKeyValueToMongoObjectKeyValue = (restKey, restValue, schema) =>
   value = mapValues(restValue, transformInteriorValue);
+
+  if (
+    typeof value === 'object' &&
+    '_bsontype' in value &&
+    !value[Symbol.for('@@mdb.bson.version')]
+  ) {
+    value[Symbol.for('@@mdb.bson.version')] = 5.0;
+  }
```
Hi @mtrezza , i work with @coratgerl , we will take a look about this seems weird to me aslo.

mtrezza replied:
Please feel free; I would be fine with merging it into Parse Server 7, i.e. we can merge it on Nov 1st, see https://github.com/parse-community/parse-server/pull/8761#discussion_r1375255686. But any help on this matter is much appreciated.

mtrezza replied:
@Moumouls @coratgerl we'll go ahead and merge this for Parse Server 7, unless you still want to investigate this bit.

Moumouls replied:
We need definitely investigate failures without this code @mtrezza before merging, until we have a clear explanation and in case of this code need to be kept a detailed comment about why we have this.

I'll try to take a look this week with @coratgerl , we will do our best !

mtrezza replied:
I agree, that would he great.

Moumouls commented on package-lock.json:
we should may be restore package lock from master. Too many change

coratgerl replied:
Hey @Moumouls ✋,
The package lock changes is normal since we have changes on the used libraries no ?

mtrezza replied:
To make sure we have only the changes needed, you could use the lock file from `alpha` and only run `npm i -E mongodb@5.9.0` and `npm i -ED mongodb-runner@5.4.4`. Run this in a npm environment that uses lock file v2, e.g. the npm that comes with Node 16.

coratgerl replied:
It's done thanks !

mtrezza replied:
Lock changes LGTM

coratgerl committed:
fix: package lock

mtrezza committed:
Merge branch 'alpha' into feat-support-mongo-v7

mtrezza committed:
Merge branch 'alpha' into feat-support-mongo-v7

coratgerl committed:
fix: remove the bsontype

coratgerl committed:
fix: remove test

mtrezza commented on spec/ParseQuery.Aggregate.spec.js:
```diff
@@ -845,6 +845,26 @@ describe('Parse.Query Aggregate testing', () => {
       });
   });
 
+  it('match date query - createdAt with bsontype', done => {
```
I'm not sure this tests is correct; the title implies "with bsontype", but there is no explicit bsontype being used, it's just a date string?
What's the purpose of `const now` and `const today`, why not just write `new Date().toISOString()`?

mtrezza committed:
Merge branch 'alpha' into feat-support-mongo-v7

mtrezza commented:
I'm a bit confused by the changes here; in https://github.com/parse-community/parse-server/pull/8761/commits/34015d5aad517237f3a0c8043685fe66b65d7f8b this has been removed:

<img width="1146" alt="image" src="https://github.com/parse-community/parse-server/assets/5673677/ce3df4d8-1a7a-424b-8c93-5fc1bec8c5e8">

But when I look at the the changes overview, I only see this:

<img width="1164" alt="image" src="https://github.com/parse-community/parse-server/assets/5673677/41f10484-4ccc-4ed8-9ac3-c4f17550628a">

mtrezza reviewed (approved):
This looks good, is this ready for merge? I've marked this as breaking, since BSON type support has been removed, even though it was never explicitly documented. Maybe we should still mention it in the changelog, how about:

> Support for BSON type "code" in `Parse.Query` is removed.

kantorcodes commented:
is there a tag to use mongo 7 support yet?

parseplatformorg commented:
🎉 This change has been released in version [7.0.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/7.0.0-alpha.1)

abhisheksharma010 reviewed (approved):
Good

parseplatformorg commented:
🎉 This change has been released in version [7.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/7.0.0-beta.1)

parseplatformorg commented:
🎉 This change has been released in version [7.0.0](https://github.com/parse-community/parse-server/releases/tag/7.0.0)
